messageBundlesCompiler.js

```diff
@@ -0,0 +1,198 @@
+
+var fs = require("fs");
+var path = require("path");
+
+var DEFAULT_FILE_TYPE = "json",
+    DEFAULT_PARSER = JSON;
+
+/**
+ * Generates a list of the least to the most specific locales that can be used.
+ * @param locale {String} A locale, e.g. "en_us".
+ * @param defaultLocale {String} The locale to be used in case messages from all
+ * previous locales are missing.
+ * @returns {String[]} A list of locales. For example, for the "fr_FR" locale and
+ * in case the default locale is "en", the function would return ["en", "fr", "fr_FR"].
+ */
+function getLocaleVersions(locale, defaultLocale) {
+    var segments = locale.split("_");
+
+    var versions = segments.map(function (segment, index) {
+        return segments.slice(0, index + 1).join("_");
+    });
+
+    if (defaultLocale) {
+        versions.unshift(defaultLocale);
+    }
+
+    return versions;
+}
+
+/**
+ * In case there are only partial messages for a given locale, this function adds
+ * the missing messages using the messages for all versions of the locale in the
+ * order in which they are returned by the `getLocaleVersions` function.
+ * @param messageBundles {Object} A raw map of collected bundles. It is of type
+ * { locale1: { key1: message1, ... }, locale2: { key2: message2 }, ... }
+ * @param defaultLocale {String} The default locale to be used.
+ * @returns {Object} The enhanced message bundles object.
+ */
+function includeFallbackOptions(messageBundles, defaultLocale) {
+    var result = {};
+
+    Object.keys(messageBundles).forEach(function (locale) {
+        var localeVersions = getLocaleVersions(locale, defaultLocale),
+            messageBundle = {};
+
+        localeVersions.forEach(function (localeVersion) {
+            if (messageBundles[localeVersion]) {
+                messageBundle = Object.assign(messageBundle, messageBundles[localeVersion]);
+            }
+        });
+
+        result[locale] = messageBundle;
+    });
+
+    return result;
+}
+
+/**
+ * Collects all file names of the specified type from the given directory.
+ * @param dir {String} The name of the directory from which files are to
+ * be collected.
+ * @param fileType {String} The file type that is to be searched for.
+ * @returns {String[]} The list of matching file names.
+ */
+function collectFilesByType(dir, fileType) {
+    var typeRegex = "\\." + fileType + "$";
+
+    var filePaths = fs.readdirSync(dir)
+        .filter(function matchesFileType(filename) {
+            return filename.match(typeRegex);
+        }).map(function (filename) {
+            return path.join(dir, filename);
+        });
+
+    return filePaths;
+};
+
+/**
+ * Loads synchronously all message bundles from the passed directories.
+ * @param bundlesDirs {String[]} The list of directrories' names containing message bundles.
+ * @param fileType {String} The file type to be searched for in the given folders.
+ * @param parser {Object} A parser object that has a `parse` method. An example of such is `JSON`.
```

Comment:
Rather than accepting a single parser, it would be more helpful to accept a map of file extensions to parsers - e.g. {"json": JSON,
 "json5": JSON5
}